:memo: Rework environments overview (#2894)
platformsh/platformsh-docs: docs/src/environments/_index.md

---
title: "Manage Platform.sh environments"
weight: -75
layout: single
sidebarTitle: Manage environments
description: Learn what environments on Platform.sh are and how to take advantage of them.
---

A Platform.sh environment contains one instance of an app (or [group of apps](../create-apps/multi-app.md))
with all the services needed for it to run.

Each project can include multiple environments,
often divided into [environment types](../administration/users.md#environment-types).
If you have a live site, you have at least a production environment.

> You can think of other environments as copies of your live site,
> where you can run tests without worrying about damaging anything in production.

Reviewer: :clipboard: 
```suggestion
You can think of other environments as copies of your live site
where you can run tests without worrying about damaging anything in production.
```

Once you have completed your tests,
you can merge your changes so that they're instantly and seamlessly deployed to production.

> You may create environments for development, testing, staging, review, and so on.

Reviewer: :clipboard: Since you pointed [this out to me](https://developers.google.com/style/word-list#may)
```suggestion
You can create environments for development, testing, staging, review, and so on.
```

To organize your environments, you can create [hierarchical relationships](#hierarchy).

## Create environments

> To create new environments, you can:

Reviewer: :hammer_and_wrench: 
```suggestion
You can create new environments in one of two ways:
```

> -   push a local branch through Git or a [source integration](../integrations/source/_index.md)
> 
> -   [branch](../other/glossary.md#branch) existing environments using the [CLI](/administration/cli/_index.md)
>     or the [Console](../administration/web/_index.md)

Reviewer: :clipboard: 
```suggestion
-   Push a local branch through Git or a [source integration](../integrations/source/_index.md).

-   [Branch](../other/glossary.md#branch) existing environments using the [CLI](/administration/cli/_index.md)
    or the [Console](../administration/web/_index.md).
```

When you branch an environment, you might want to create exact replicas of it.
In this case, each new environment inherits all of the data and services from its parent environment.
This includes databases, network storage, queues, and routing configurations.

You can create Platform.sh environments on demand.
Each environment is tied to a Git branch.

> If you use a [source integration](../integrations/source/_index.md),

Reviewer: :pick: Since source integrations are linked just before this in this section, it might not be necessary here. :shrug: 
```suggestion
If you use a source integration,
```

-   [Inactive](../other/glossary.md#inactive-environment):
    An environment that isn't deployed and has no services or data, only code.

You can see the status of your environments in the [Console](../administration/web/_index.md) or the [CLI](/administration/cli/_index.md).

{{< codetabs >}}
+++
title=In the Console
+++

When you open your project, inactive environments appear lighter in the environment list.
You also can't select them from the **Environment** dropdown menu.

To check the status of an environment,
you can also open it and view its information panel.

<--->
+++
title=Using the CLI
+++

To check the status of all your environments, from your project directory run the `platform environments` command.

{{< /codetabs >}}

You can [change an environment's status](./deactivate-environment.md) at any time.

## Organize your environments

### Hierarchy

![Environment hierarchy](/images/management-console/environments.png "0.5")

In Platform.sh, your environments are organized in a hierarchy featuring parent and child environments.

When you [branch](../other/glossary.md#branch) an environment,
the parent of the new environment is the environment it was created from.

> You can [change the environment's parent](./change-parent.md) after it's been created using the [CLI](/administration/cli/_index.md).

Reviewer: :hammer_and_wrench: This makes it seem like it's only possible with the CLI, while the link shows how to do it in the Console.
```suggestion
You can [change the environment's parent](./change-parent.md) after it's been created.
```

When you push a branch through Git or a [source integration](../integrations/source/_index.md),

> the parent environment of the new environment is your [default environment](#default-environment),
> unless you specify otherwise using the `--parent` flag in the [CLI](/administration/cli/_index.md).

Reviewer: :pick: You convinced me yesterday that I was wrong and it'd be better not to include this here. :see_no_evil:  Based on the way it's pushed through the start of the sentence. I think I just read it too quickly and you were right before. But this information is nice to have somewhere, so it doesn't hurt if you want to include it after figuring it out for me.

```suggestion
the parent environment of the new environment is your [default environment](#default-environment).
```

You can decide which structure best fits your needs. 

#### Possible approaches

You may want to take one of the following approaches when creating your workflows:

-   **Agile**: A child environment per sprint.
    Each story in the sprint can have its own environment as a child of the sprint environment.

-   **Developer-centric**: One QA environment and a few development environments
    (*per developer*, *per task*, or similar).

-   **Testing**: An operational test environment, a user test environment, and a few unit test environments.

-   **Hotfix**: One environment for each bug, security issue, or hotfix that needs deployment.

#### Example workflow

Example Agile workflow:

1.  An admin [branches](../other/glossary.md#branch) the Live (default) environment to create a Sprint environment.

2.  The admin gives each developer permission to branch the Sprint environment to create new feature environments.

    ![Agile branches with Live as the top parent, Sprint as a child, and Feature 1 and Feature 2 as children of Sprint](/images/workflow/branches.svg "0.2")

3.  Feature 1 is developed and work is reviewed by accessing the deployed Feature 1 environment.

> 4.  When the review is done, Feature 1 is [merged](../other/glossary.md#merge) into the Sprint environment.

Reviewer: :nail_care: I still feel like an image of the merge would help make it feel consistent.

---
title: "Manage Platform.sh environments"
weight: -75
layout: single
sidebarTitle: Manage environments
description: Learn what environments on Platform.sh are and how to take advantage of them.
---

A Platform.sh environment contains one instance of an app (or [group of apps](../create-apps/multi-app.md))
with all the services needed for it to run.

You can think of an environment as a complete working website
that's related to but safely isolated from others in your project.
So you can run tests and review a complete working copy without worrying about damaging anything in production.

Each project includes multiple environments,
often divided into [environment types](../administration/users.md#environment-types).
If you have a live site, you have at least a production environment.
You may also have additional environments for development, testing, staging, review, and so on.

> To organize your environments, you can create [hierarchical relationships](#hierarchy).

Reviewer: :+1: I like the move here. Makes sense right after listing how many there can be.

To create new environments, you can branch existing environments using the [CLI](/administration/cli/_index.md)
or the [Console](../administration/web/_index.md).
Each created environment is an exact replica of its parent environment.
This means new environments inherit all of the data and services from their parent environment.
This includes databases, network storage, queues, and routing configurations.

You can create environments on demand.
Each environment is tied to a Git branch.
With [Bitbucket](../integrations/source/bitbucket.md) and [GitHub](../integrations/source/github.md) integrations,
you can even get a "development server" automatically for every pull request.

You can also have branches that aren't tied to a running instance of your application.
These are called [inactive environments](#environment-status).

## Default environment

Your default environment comes from your default branch and is a [production environment](../administration/users.md#environment-types).
Your project must have a default environment,
but you can [name it as you want](./default-environment.md).

If you subscribed to a production plan, this environment is your **live site**.

> You can give it a [custom domain name](../domains/steps/_index.md) and a [custom TLS certificate](../domains/steps/tls.md).

Reviewer: :question: I'm not sure if the certificates are important enough to the idea of environments to merit mentioning here. The domains seem clear.

## Environment status

Your environments can have one of two statuses:

-   [Active](../other/glossary.md#active-environment):
    A deployed environment with services and data.

-   [Inactive](../other/glossary.md#inactive-environment):
    An environment that isn't deployed and has no services or data, only code.
    For example, when you push a local branch created with Git to your project,

>     it creates an inactive environment.

Reviewer: :lady_beetle: Technically, there's an option to push with the CLI branches created with Git and activate them on the push. I'd also say the `it` is slightly vague.

You can see the status of your environments in the [Console](../administration/web/_index.md) or the [CLI](/administration/cli/_index.md).

{{< codetabs >}}
+++
title=In the Console
+++

When you open your project, inactive environments appear lighter in the environment list.
You also can't select them from the **Environment** dropdown menu.

To check the status of an environment,
you can also open it and view its information panel.

<--->
+++
title=Using the CLI
+++

> To check the status of all your environments, run the `platform environments` command within your project directory.

Reviewer: :pick: 
```suggestion
To check the status of all your environments, from your project directory run the `platform environments` command.
```

{{< /codetabs >}}

> Note that you can [change an environment's status](./deactivate-environment.md).

Reviewer: :hammer_and_wrench: I wouldn't say that this is a side note, really.
```suggestion
You can [change an environment's status](./deactivate-environment.md).
```

## Organize your environments

### Hierarchy

> ![Environment hierarchy](/images/management-console/environments.png "0.5")

Reviewer: :lady_beetle: This isn't exactly what I get when I have active child environments, but I'm not sure if it's important or a bug or what. Instead of blue branch markers, I see nothing.

Each child environment can [sync](../other/glossary.md#sync) code and/or data down from its parent
and [merge](../other/glossary.md#merge) code up to its parent.
You can use child environments for development, staging, and testing.

When you [branch](../other/glossary.md#branch) an environment to create a new child environment,
its parent is the environment it was created from.
If you push a branch through Git or a [source integration](../integrations/source/_index.md),

> the parent environment is your [default environment](#default-environment).

Reviewer: :lady_beetle: Might be worth mentioning what I think the original was trying to get at (but not really succeeding): that you can set different parents through the CLI when branching/pushing.

You can always [change an environment's parent](./change-parent.md).

### Workflows

Since you can customize your [environment hierarchy](#hierarchy), you can create your own workflows. 

> You don't have to follow any rules when branching environments.

Reviewer: :hammer_and_wrench: I think your wording is clearer than the original, but I'm not really sure the sentence adds much. Unless you seem something useful I'm missing.
```suggestion
```

You can decide which structure best fits your needs. 

> #### Possible approaches

Reviewer: :+1: I like the added structure

You may want to take one of the following approaches when creating your workflows:

-   **Agile**: A child environment per sprint.
    Each story in the sprint can have its own environment as a child of the sprint environment.

-   **Developer-centric**: One QA environment and a few development environments
    (*per developer*, *per task*, or similar).

-   **Testing**: An operational test environment, a user test environment, and a few unit test environments.

> -   **Hotfix**: One environment for every bug, security, or hotfix that needs deployment.

Reviewer: :hammer_and_wrench: This is trying to get a parallel with `fix` with each part in the list, but it doesn't really work when `hotfix` is a single word.
```suggestion
-   **Hotfix**: One environment for each bug, security issue, or hotfix that needs deployment.
```

#### Example workflow

> Example of an Agile workflow :

Reviewer: :hammer_and_wrench: 
```suggestion
Example Agile workflow:
```

> 1.  The admin [branches](../other/glossary.md#branch) the Live environment to create a Sprint environment.

Reviewer: :hammer_and_wrench: 
```suggestion
1.  An admin [branches](../other/glossary.md#branch) the Live (default) environment to create a Sprint environment.
```

2.  The admin gives each developer permission to branch the Sprint environment to create new feature environments.

    ![Agile branches with Live as the top parent, Sprint as a child, and Feature 1 and Feature 2 as children of Sprint](/images/workflow/branches.svg "0.2")

> 3.  Feature 1 is developed and work is reviewed by accessing the deployed Feature 1 environment.

Reviewer: :lady_beetle: It seems like an image of this step would be nice, given that all the others have it.

5.  The remaining features [sync](../other/glossary.md#sync) with the Sprint environment.
    This ensures their working environment is up-to-date with the latest code.

    ![Features from the Sprint environment are synced with the Feature 2 environment](/images/workflow/sync.svg "0.2")

6.  When the sprint is complete and all features merged into the Sprint environment,
    the admin makes a backup of the live site.

7.  The admin merges the Sprint environment into the Live environment.

    ![Features from the Sprint environment are merged into the Live environment](/images/workflow/merge-live.svg "0.2")

8.  The admin syncs the next sprint's environment with the Live environment
    to repeat and continue the development process.

### Naming conventions

You can organize and work with your development environments in many different ways.
It can help to introduce a convention for how you name and structure your environments.

For each environment, choose a name that represents what the environment is for.
If you use Agile, for example, you can adopt a naming convention similar to the following:

```text
Live

>  Sprint1
>    Feature1
>    Feature2
>    Feature3
>  Sprint2
>    Feature1
>    Feature2

Reviewer: :clipboard: A single space at the start of a line isn't great for being able to see the difference. I'd always have at least 2 for indentation
```suggestion
  Sprint1
    Feature1
    Feature2
    Feature3
  Sprint2
    Feature1
    Feature2
```

```

If you prefer splitting your environments per developer and having a specific environment for each task or ticket,
you can adopt a naming convention similar to the following:

```text
Staging

>  Developer1
>    Ticket-526
>    Ticket-593
>  Developer2
>    Ticket-395

Reviewer: :clipboard: 
```suggestion
  Developer1
    Ticket-526
    Ticket-593
  Developer2
    Ticket-395
```
